mobilecoind-dev-faucet/src/slam/prepared_utxo.rs

```diff
@@ -0,0 +1,214 @@
+                }
+            }
+            tries += 1;
+        };
+
+        // The Tx builder sorts the ring anyways, so it doesn't matter if we always put
+        // the real input first.
+        let (ring, membership_proofs): (Vec<TxOut>, Vec<TxOutMembershipProof>) = proofs_resp
+            .get_output_list()
+            .iter()
+            .chain(mixins_resp.get_mixins().iter())
+            .map(|tx_out_with_proof| {
+                Ok((
+                    tx_out_with_proof.get_output().try_into()?,
+                    tx_out_with_proof.get_proof().try_into()?,
+                ))
+            })
+            .collect::<Result<Vec<_>, ConversionError>>()
+            .map_err(|err| format!("Conversion error: {}", err))?
+            .into_iter()
+            .unzip();
+
+        Ok(PreparedUtxo {
+            index,
+            utxo_record,
+            ring,
+            membership_proofs,
+            real_input_index: 0,
+        })
+    }
+
+    async fn get_proofs_and_mixins(
+        utxo_record: &UtxoRecord,
+        mobilecoind_api_client: &MobilecoindApiClient,
+    ) -> Result<
+        (
+            mc_mobilecoind_api::GetMembershipProofsResponse,
+            mc_mobilecoind_api::GetMixinsResponse,
+        ),
+        String,
+    > {
+        // Get a membership proof for this utxo
+        let mut req = mc_mobilecoind_api::GetMembershipProofsRequest::new();
+        req.set_outputs(RepeatedField::from(vec![utxo_record
+            .utxo
+            .get_tx_out()
+            .clone()]));
+
+        let proofs_resp = mobilecoind_api_client
+            .get_membership_proofs_async(&req)
+            .map_err(|err| format!("Failed to request membership proofs: {}", err))?
+            .await
+            .map_err(|err| format!("Request membership proofs ended in error: {}", err))?;
+
+        // Get mixins for this utxo
+        let mut req = mc_mobilecoind_api::GetMixinsRequest::new();
+        req.set_num_mixins(RING_SIZE as u64 - 1);
+        req.set_excluded(RepeatedField::from(vec![utxo_record
+            .utxo
+            .get_tx_out()
+            .clone()]));
+
+        let mixins_resp = mobilecoind_api_client
+            .get_mixins_async(&req)
+            .map_err(|err| format!("Failed to request mixins: {}", err))?
+            .await
+            .map_err(|err| format!("Request mixins ended in error: {}", err))?;
+
+        Ok((proofs_resp, mixins_resp))
+    }
+
+    /// Build a Tx from a prepared utxo
+    pub fn build_tx(
+        &self,
+        tombstone_block: u64,
+        recipient: &PublicAddress,
+        account_key: &AccountKey,
+        network_state: &GetNetworkStatusResponse,
+    ) -> Result<Tx, String> {
+        let mut rng = thread_rng();
+        // Get block version to target
+        let block_version =
+            BlockVersion::try_from(network_state.get_last_block_info().network_block_version)
+                .map_err(|err| format!("Block version: {}", err))?;
```

Comment:
It seems useful to include the invalid block version, and specify that it came from the network state.
```suggestion
        let block_version = network_state.get_last_block_info().network_block_version;
        let block_version = BlockVersion::try_from(block_version)
            .map_err(|err| format!("Got invalid block version {} from network ({})", block_version, err))?;
```